sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeFileAsyncClient.java

```diff
@@ -148,6 +183,368 @@ public String getFileName() {
+        } catch (IOException e) {
+            throw logger.logExceptionAsError(new UncheckedIOException(e));
+        } finally {
+            uploadFileCleanup(channel);
+        }
+
+        return retVal;
+    }
+
+    private Mono<Void> uploadBlocks(long fileOffset, long fileSize, ParallelTransferOptions parallelTransferOptions,
+        Integer originalBlockSize, PathHttpHeaders headers, DataLakeRequestConditions requestConditions,
+        AsynchronousFileChannel channel) {
+        // parallelTransferOptions are finalized in the calling method.
+
+        // See ProgressReporter for an explanation on why this lock is necessary and why we use AtomicLong.
+        AtomicLong totalProgress = new AtomicLong();
+        Lock progressLock = new ReentrantLock();
+
+        return Flux.fromIterable(sliceFile(fileSize, originalBlockSize, parallelTransferOptions.getBlockSize()))
+            .flatMap(chunk -> {
+                Flux<ByteBuffer> progressData = ProgressReporter.addParallelProgressReporting(
+                    FluxUtil.readFile(channel, chunk.getOffset(), chunk.getCount()),
+                    parallelTransferOptions.getProgressReceiver(), progressLock, totalProgress);
+
+                return appendWithResponse(progressData, fileOffset + chunk.getOffset(), chunk.getCount(), null,
+                    requestConditions.getLeaseId());
+            })
+            .then(Mono.defer(() ->
+                flushWithResponse(fileSize, false, false, headers, requestConditions)))
+            .then();
+    }
+
+    /**
+     * RESERVED FOR INTERNAL USE.
+     *
+     * Resource Supplier for UploadFile.
+     *
+     * @param filePath The path for the file
+     * @return {@code AsynchronousFileChannel}
+     * @throws UncheckedIOException an input output exception.
+     */
+    protected AsynchronousFileChannel uploadFileResourceSupplier(String filePath) {
+        try {
+            return AsynchronousFileChannel.open(Paths.get(filePath), StandardOpenOption.READ);
+        } catch (IOException e) {
+            throw logger.logExceptionAsError(new UncheckedIOException(e));
+        }
+    }
+
+    private void uploadFileCleanup(AsynchronousFileChannel channel) {
+        try {
+            channel.close();
+        } catch (IOException e) {
+            throw logger.logExceptionAsError(new UncheckedIOException(e));
+        }
+    }
+
+    private List<FileRange> sliceFile(long fileSize, Integer originalBlockSize, int blockSize) {
+        List<FileRange> ranges = new ArrayList<>();
+        if (fileSize > 100 * Constants.MB && originalBlockSize == null) {
+            blockSize = BlobAsyncClient.BLOB_DEFAULT_HTBB_UPLOAD_BLOCK_SIZE;
+        }
+        for (long pos = 0; pos < fileSize; pos += blockSize) {
+            long count = blockSize;
+            if (pos + count > fileSize) {
+                count = fileSize - pos;
+            }
+            ranges.add(new FileRange(pos, count));
+        }
+        return ranges;
```

Comment:
These I think can also move into the common utility, right?

Reply:
yep, there was some difficulty with some of the logging stuff but it should go in common - I'll work on that